lib/Library.php

```diff
@@ -140,31 +140,49 @@ public static function GetEbooksByTag(string $tag): array{
 	}
 
 	/**
-	 * @return array<string, Collection>
+	 * @return array<Collection>
 	 * @throws Exceptions\AppException
 	 */
 	public static function GetEbookCollections(): array{
-		/** @var array<string, Collection> */
-		return self::GetFromApcu('collections');
+		$collections = Db::Query('
+					SELECT distinct UrlName, Name
+					from Collections
+				', [], Collection::class);
+
+		$collator = Collator::create('en_US');
+		if($collator === null){
+			throw new Exceptions\AppException('Couldn\'t create collator object when getting collections.');
+		}
+		usort($collections, function($a, $b) use($collator){ return $collator->compare($a->GetSortedName(), $b->GetSortedName()); });
+		return $collections;
 	}
 
 	/**
 	 * @return array<Ebook>
-	 * @throws Exceptions\AppException
 	 */
 	public static function GetEbooksByCollection(string $collection): array{
-		// Do we have the tag's ebooks cached?
-		/** @var array<Ebook> */
-		return self::GetFromApcu('collection-' . $collection);
+		$ebooks = Db::Query('
+				SELECT e.*
+				from Ebooks e inner join Collections c using (EbookId)
+				where c.UrlName = ?
+				order by c.SequenceNumber, e.Title
+				', [$collection], Ebook::class);
+
+		return $ebooks;
 	}
 
 	/**
-	 * @return array<Tag>
-	 * @throws Exceptions\AppException
+	 * @return array<EbookTag>
 	 */
 	public static function GetTags(): array{
-		/** @var array<Tag> */
-		return self::GetFromApcu('tags');
+		$tags = Db::Query('
+				SELECT distinct t.*
```

Comment:
Again is `distinct` required? In what case would we have more than one of the same tag attached to an ebook?

Reply:
This one is a different problem: We are sharing the `Tags` table with `EbookTags` and `ArtworkTags`. If we weren't doing that, we could just select all the `Tags` and be set. Instead, we have to join with the `EbookTags` table to get the tags that are used for ebooks. When we do that join, we get records for each (ebook, tag) pair. Here's what the query looks like without `distinct`:

```
> select t.* from Tags t inner join EbookTags et using (TagId) order by t.Name limit 100;
+-------+-----------+
| TagId | Name      |
+-------+-----------+
|     7 | Adventure |
|     7 | Adventure |
|     7 | Adventure |
|     7 | Adventure |
|     7 | Adventure |
|     7 | Adventure |
|     7 | Adventure |
|     7 | Adventure |
|     7 | Adventure |
[...]
|     7 | Adventure     |
|     7 | Adventure     |
|     7 | Adventure     |
|     7 | Adventure     |
|     7 | Adventure     |
|     7 | Adventure     |
|    11 | Autobiography |
|    11 | Autobiography |
|    11 | Autobiography |
|    11 | Autobiography |
|    11 | Autobiography |
|    11 | Autobiography |
|    11 | Autobiography |
|    11 | Autobiography |
+-------+---------------+
``` 

I considered `GROUP BY` instead of `DISTINCT`, but the explain plan was the same either way:

```
> explain select distinct t.* from Tags t inner join EbookTags et using (TagId) order by t.Name;
+------+-------------+-------+--------+---------------+-----------+---------+-------------+------+----------------------------------------------+
| id   | select_type | table | type   | possible_keys | key       | key_len | ref         | rows | Extra                                        |
+------+-------------+-------+--------+---------------+-----------+---------+-------------+------+----------------------------------------------+
|    1 | SIMPLE      | et    | index  | NULL          | idxUnique | 8       | NULL        | 1314 | Using index; Using temporary; Using filesort |
|    1 | SIMPLE      | t     | eq_ref | PRIMARY       | PRIMARY   | 4       | se.et.TagId | 1    |                                              |
+------+-------------+-------+--------+---------------+-----------+---------+-------------+------+----------------------------------------------+

> explain select t.Name from Tags t inner join EbookTags et using (TagId) group by t.Name order by t.Name;
+------+-------------+-------+--------+---------------+-----------+---------+-------------+------+----------------------------------------------+
| id   | select_type | table | type   | possible_keys | key       | key_len | ref         | rows | Extra                                        |
+------+-------------+-------+--------+---------------+-----------+---------+-------------+------+----------------------------------------------+
|    1 | SIMPLE      | et    | index  | NULL          | idxUnique | 8       | NULL        | 1314 | Using index; Using temporary; Using filesort |
|    1 | SIMPLE      | t     | eq_ref | PRIMARY       | PRIMARY   | 4       | se.et.TagId | 1    |                                              |
+------+-------------+-------+--------+---------------+-----------+---------+-------------+------+----------------------------------------------+

```

Reply:
Oh, and sorry if this wasn't clear: The purpose of this `Library::GetTags()` is to get all the tags so that they can be shown on the search form. It's not to get the tags for a single ebook, which is `Ebook::GetTags()`:

https://github.com/standardebooks/web/blob/f33f219ae57bfcf50fe5d8f329f82fa3e36881c4/lib/Ebook.php#L138

I could rename this one from `Library::GetTags()` to `Library::GetAllEbookTags()` if that would help.

Reply:
OK. Let's add a `Type` column to the `Tags` table, which will be `enum('ebook','artwork')`. Then we can select on that column and things will be a lot easier. It's unlikely artwork and ebook tags will ever conflict anyway.